packages/react-instantsearch-nextjs/src/InstantSearchNext.tsx

```diff
@@ -51,11 +48,17 @@ export function InstantSearchNext<
 }: InstantSearchNextProps<TUiState, TRouteState>) {
   const isMounting = useRef(true);
   const isServer = typeof window === 'undefined';
-  const pathname = usePathname();
-  const hasRouteChanged =
-    !isServer &&
-    window[InstantSearchLastPath] &&
-    window[InstantSearchLastPath] !== pathname;
+
+  const hasRouteChanged = useMemo(() => {
+    // On server, always return false
+    if (isServer) {
+      return false;
+    }
+
+    // On client, route has changed if initialResults have been cleaned up
+    const hasInitialResults = window[InstantSearchInitialResults] !== undefined;
+    return !hasInitialResults;
+  }, [isServer]);
```

Comment:
Why a `useMemo` when `isServer` won't ever change ?

There's more overhead in calling that than just a single-line :
```ts
const hasRouteChanged = !isServer && !window[InstantSearchInitialResults
```

We can still make it multiple lines to make it more readable

Reply:
`useMemo()` prevents the value from being true during a rerender triggered by a wrapping component.

Because of strict mode, the initial results are cleaned during the automatic unmount/remount, which will cause the issue the previous PR aimed to fix.

Reply:
Ah yes because `window[InstantSearchInitialResults]` is not in the dependency array